### DISCUSSION TITLE
Welcome new users!

### DISCUSSION THREAD

**Discussion by friendly-bits** [friendly-bits/geoip-shell#4] [General]
Probably barely anybody reads this but for those who do: Lately I've been noticing that this repository is starting to attract some (positive) attention and I see that people are downloading and using this project, which I'm happy about. You are always welcome to ask questions or to ask for additional features (or report bugs if any).

At this occasion I'll also mention that the next release should be coming soon'ish, which features support for additional shells (ksh93 and yash), more robust compatibility checks, improved reliability and some security improvements as well. Keep posted.

**Comment by OIITCONZ** [comment]
Hi @friendly-bits 
Thanks for the great script. I have just tried it on an Ubuntu LTS 24.04 and it was very smooth. 

Two questions : I notice ICMP is not working. I geo white listed NZ (so banning all other countries) I cannot ping the server. Is that this shell or standard server functionality?  I can ssh in on a different to standard port so it handled that very well. 

I am wanting to ban all traffic from all countries except NZ and then either forward a set of ports or forward everything from this server to a fixed ip address. The end result is to setup a community minecraft server - which is getting hammered by scum from other countries. Keeping it NZ only keeps the ND kids we run the server for safe.  Are there any configurations that allow this? 

Thanks for the great work,

**Reply by friendly-bits** [reply]
Hi there @OIITCONZ 
I'll answer your first question first.
geoip-shell doesn't specifically block any ports by default. So probably ICMP not working is unrelated. You can check if geoip-shell has anything to do with it by issuing this command: `geoip-shell off` and checking if ping is working now. This disables geoip-shell completely. To re-enable later, use `geoip-shell on`.

**Reply by friendly-bits** [reply]
As to your second question, I'm not sure what you mean by "forward a set of ports or forward everything" but it sounds like you want to set up a tunnel. geoip-shell doesn't do tunnels but there is plenty of software for this. geoip-shell will not interfere with that software, unless for some reason that software needs to receive traffic from blocked countries. What geoip-shell does is it fetches a list of ip subnets which belong to certain country or countries (in your case NZ), creates an ipset (if using iptables) or nftables set (if using nftables) out of that list, and (in whitelist mode) creates iptables or nftables rules which block all inbound traffic which didn't come from ip addresses included in the set. When installed for all network interfaces rather than a specific WAN network interface, geoip-shell also creates some allow rules in order to not block DHCPv6 and other essential communication (which is non-routable, so should not affect the effectiveness of geoip blocking).

Generally, if you need things which should occasionally receive traffic from blocked countries, you would have to either add those countries to the whitelist, or come up with some sort of a workaround, for example here users came up with a workaround to allow letsencrypt to renew certificates (you will probably need to use Google automatic translation):
https://forum.netcup.de/administration-of-a-server-vserver/vserver-server-kvm-server/18634-geoblocking/

If you have any additional questions about geoip-shell, please feel free to ask.

**Reply by friendly-bits** [reply]
Also note that the available geoip data is not 100% reliable. In some cases some ip subnets may be listed under one country but actually belong to a different country. This is not frequent but this happens. This is a shortcoming of all geoip blocking, not just geoip-shell. Specifically for NZ, I don't know how many ip subnets are assigned wrong and in which way (either an address in NZ is listed under a different country or an address in a different country appears under NZ). It could be that NZ is not affected at all, or it could be that you will have issues. You will just need to try and see.

geoip blocking is very effective for reducing the attack surface despite this shortcoming, however if you need a guarantee that 100% percent of people trying to access your server from NZ are not geoblocked, this guarantee can not be given.

**Reply by friendly-bits** [reply]
> geoip-shell will not interfere with that software, unless for some reason that software needs to receive traffic from blocked countries.

To clarify this point, geoip-shell creates a firewall rule which will allow the locally running software to make outbound connections to blocked countries and receive response. Only if the software is listening for inbound connection requests (without establishing an outbound connection first) then it will not be able to see it if given request comes from a blocked country.

**Comment by OIITCONZ** [comment]
Basically I am wanting to block all ips from anywhere but NZ.

Then set up rules similar to this :
sudo iptables -A INPUT -m set --match-set nz_ips src -j ACCEPT
sudo iptables -t nat -A PREROUTING -m set --match-set nz_ips src -j DNAT 
--to-destination 123.456.789.1
sudo iptables -A FORWARD -d 123.456.789.1 -j ACCEPT

Obviously 123.456.789.1 is not my real destination ip address.

My concern is if i do rule 1 above - I cannot set to only allow NZ ips - 
but geoip should be doing that so I can ignore rule 1
Rule 2 - Is where this gets difficult. How do I set a prerouting so only 
acceptable ips from an NZ range are routed out by rule 3?
Rule 3 How and where would I set this. My server is an out of the box 
Ubuntu LTS 24.04 . I turned off UFW and turned on IPTables.

The iptables rules above are from examples for doing this using iptables 
and ipset sources but none of them seemed to work - leaving me unable to 
ping out, resolve DNS etc.
Yours lets me do those things so now I am trying to add the forwarding 
rules as the final step.

The end goal is only allow minecraft players from NZ to connect to the 
server. We are getting hammered by dipsticks from overseas which is 
chewing up bandwidth and CPU faster than it should for a small server.
Ideally only ports 4000 - 40010 would be forwarded and all others 
dropped (SSH excluded)
This box is a relay / router to stand in front of the minecraft server.
I could run geoshell on the minecraft server but there are other 
considerations (like 11 other servers behind the same nat) and some of 
them need overseas to connect.

I suspect the best option is put MC on its own box, and run geoip on the 
MC box. I would prefer to use a VPS with a wide bandwidth and other 
firewalls infront rather than direct to my server.

Thanks for your help and answer.

Shane










Shane Hollis
Outsourced Information Technology Limited
Mob: 022 626 5084

-- ----
All Outsourced Information Technology Limited's work, sales and 
correspondence is covered under our terms of service and privacy 
policies found at https://outsourcedit.co.nz/tandc .

Our online invoicing system is found at 
https://books.zoho.com/portal/outsourcedinformationtechnolog and 
invoices, receipts and quotes can be accessed there.

All goods and services sold are covered under the Sale of Goods Act and 
the Consumer Guarantees Act of NZ and will be liable for GST where 
applicable.


------ Original Message ------
From "friendly-bits" ***@***.***>
To "friendly-bits/geoip-shell" ***@***.***>
Cc "Secure Log On and Form System" ***@***.***>; 
"Mention" ***@***.***>
Date 24/08/2024 8:41:32 pm
Subject Re: [friendly-bits/geoip-shell] Welcome new users! (Discussion 
#4)

**Reply by friendly-bits** [reply]
Hi, did you manage to achieve the config you wanted? I posted some replies below which I'm not sure you've seen.

**Comment by friendly-bits** [comment]
First off, you should probably get acquainted with netfiler hooks:
https://wiki.nftables.org/wiki-nftables/index.php/Netfilter_hooks

(this is from the nftables wiki but the netfilter kernel infrastructure is used by iptables as well)

geoip-shell creates rules attached to the `prerouting` hook. With nftables this is done via a chain which attaches directly to the hook, with iptables geoip-shell uses the `PREROUTING` built-in chain in table `mangle` (read section 2 of [NOTES.md](https://github.com/friendly-bits/geoip-shell/blob/main/Documentation/NOTES.md) for additional details).

This means that geoip blocking is done before any routing decisions about the ingress traffic are made. In other words, traffic from geoblocked countries will not reach table nat where your second rule is located. So with geoip-shell, you don't need the `match-set` in that rule.

Now correct me if I'm wrong but it sounds like you actually want to geoblock **only** traffic arriving on some specific ports and let all the other traffic reach its destination. You can do this with geoip-shell: read [NOTES.md](https://github.com/friendly-bits/geoip-shell/blob/main/Documentation/NOTES.md), sections 9-11.

Last but not least please run `geoip-shell status` and make sure that the firewall backend is set to iptables. When nftables is present in the system, geoip-shell defaults to nftables. So if you find out that currently the firewall backend is nftables then use this command to change the firewall backend to iptables:
`geoip-shell configure -w ipt`

**Reply by friendly-bits** [reply]
You can also read this for more information on how packets traverse tables and chains with iptables:
https://www.baeldung.com/linux/iptables-chains-tables-traversal

**Reply by friendly-bits** [reply]
Sorry, for some reason the links to NOTES.md didn't work in the above answer, fixed now.

**Comment by friendly-bits** [comment]
I may not have answered all of your questions directly but I hope that with the above information, you can come up with the specific solution you need. If you still need more help after you have processed that information, feel free to ask.

**Comment by OIITCONZ** [comment]
Hi,

Yes - I saw the replies.
No - Not achieved yet. I had a client whose system went down ( gotta 
love electricians cutting cables) so had to put this on pause until this 
afternoon.

I'll let you know how it goes.
Thanks you. I see you are well named friendly-bits.

BTW : Do you have a donate link? Buy you a coffee or two?

Shane

**Reply by friendly-bits** [reply]
No need really for a coffee but you can give the repo a star if you fancy.
And please do keep me posted.

**Comment by Fantanchero** [comment]
Hello.
I'm new on sysadmin.
Trying to setup geoip it gives me the following error:
" ...... Error: Could not process rule: Numerical result out of range..........."
Maybe I'm missing or outdated some dependency?
Thanks in advance.

Debian 12.

![Captura de pantalla 2024-10-09 185938](https://github.com/user-attachments/assets/5145bded-fadb-4e3f-9fb8-4e81ac7f8abb)

**Reply by friendly-bits** [reply]
Hi, could you please post this as text rather than picture?
Did you try again?

Also what's the output of:
```
nft --version
uname -a
```

**Reply by Fantanchero** [reply]
Here is the text version of sudo sh geoip-shell-install.sh
```
fanta@linux:~/firewallsito/geoip-shell-0.5.8$ sudo sh geoip-shell-install.sh
[sudo] password for fanta:
Creating backup of geoip-shell ip sets... netlink: Error: cache initialization failed: Numerical result out of range
backup: Error: Failed to back up geoip-shell ip sets.
Deleting the data directory '/var/lib/geoip-shell'...
install: Cleaning up previous installation (if any)...
Deleting the iplist directory '/tmp/geoip-shell'...
Deleting geoip-shell main scripts from /usr/bin...
Deleting the library scripts directory '/usr/lib/geoip-shell'...
Uninstall complete.

Copying scripts to /usr/bin... Ok.
Copying library scripts to /usr/lib/geoip-shell... Ok.
Existing config file found. [K]eep previous config or [f]orget the old config? [k|f] or [a] to abort setup.
k|f|a: f


Select geoip blocking mode: [w]hitelist or [b]lacklist, or [a] to abort.
w|b|a: w

Please enter country codes to include in geoip whitelist.
Country codes (2 letters) or [a] to abort: UY

Please enter your country code.
It will be used to check if your geoip settings may block your own country and warn you if so.
Country code (2 letters)/Enter to skip: UY

Does this machine have dedicated WAN network interface(s)? [y|n] or [a] to abort.
For example, a router or a virtual private server may have it.
A machine connected to a LAN behind a router is unlikely to have it.
It is important to answer this question correctly.
y|n|a: y

*NOTE*: Geoip firewall rules will be applied to specific network interfaces of this machine.
All found network interfaces: venet0
Autodetected WAN interfaces: venet0
[c]onfirm, c[h]ange, or [a]bort?
c|h|a: c
manage: Applying config...

Checking connectivity... Ok.

Checking for ip list updates on the RIPE server...

Fetching ip list for country 'UY' from RIPE...
################################################################################################################# 100.0%Fetch successful.

Parsing ip list for 'UY_ipv4'... Ok.
Validating 'UY_ipv4'... Ok.
Validated subnets for 'UY_ipv4': 86.

Parsing ip list for 'UY_ipv6'... Ok.
Validating 'UY_ipv6'... Ok.
Validated subnets for 'UY_ipv6': 39.

Adding ip set 'UY_ipv4_2024-10-08_geoip-shell'... /dev/stdin:1:1-4: Error: Could not process rule: Numerical result out of range
add set inet geoip-shell UY_ipv4_2024-10-08_geoip-shell { type ipv4_addr; flags interval; auto-merge; policy performance; elements={ 45.6.248.0/22,45.6.252.0/22,45.228.64.0/23,69.164.80.0/24,69.164.91.0/24,69.164.98.0/24,131.0.212.0/22,138.59.12.0/22,138.99.44.0/22,152.156.0.0/16,161.0.120.0/21,164.73.0.0/16,167.56.0.0/13,167.108.0.0/16,167.116.0.0/16,168.121.184.0/22,170.247.168.0/22,179.0.156.0/22,179.24.0.0/13,181.103.64.0/20,181.224.164.0/23,186.8.0.0/18,186.8.64.0/18,186.8.128.0/17,186.48.0.0/14,186.52.0.0/14,190.0.128.0/19,190.64.0.0/17,190.64.128.0/17,190.99.96.0/20,190.108.0.0/19,190.112.52.0/22,190.112.192.0/22,190.112.200.0/22,190.132.0.0/16,190.133.0.0/16,190.134.0.0/15,192.146.237.0/24,200.0.68.0/23,200.0.86.0/23,200.0.88.0/24,200.0.204.0/22,200.2.32.0/19,200.3.12.0/22,200.7.84.0/23,200.7.86.0/23,200.10.60.0/23,200.10.62.0/23,200.12.155.0/24,200.12.156.0/24,200.35.152.0/22,200.40.0.0/17,200.40.128.0/18,200.40.192.0/19,200.40.224.0/20,200.40.240.0/20,200.58.128.0/20,200.58.144.0/20,200.71.0.0/20,200.71.16.0/20,200.93.240.0/24,200.93.241.0/24,200.93.242.0/24,200.93.243.0/24,200.93.244.0/24,200.93.245.0/24,200.93.246.0/24,200.93.247.0/24,200.108.192.0/19,200.108.224.0/20,200.108.240.0/20,200.115.72.0/21,200.124.192.0/20,200.124.208.0/20,200.125.0.0/20,200.125.16.0/20,200.125.32.0/20,200.125.48.0/20,201.217.128.0/19,201.217.160.0/19,201.219.224.0/21,201.219.252.0/22,201.221.0.0/20,201.221.16.0/20,201.221.32.0/20,201.221.48.0/20,}
^^^^
/dev/stdin:1:57-1: Error: Could not process rule: No such file or directory
add set inet geoip-shell UY_ipv4_2024-10-08_geoip-shell { type ipv4_addr; flags interval; auto-merge; policy performance; elements={ 45.6.248.0/22,45.6.252.0/22,45.228.64.0/23,69.164.80.0/24,69.164.91.0/24,69.164.98.0/24,131.0.212.0/22,138.59.12.0/22,138.99.44.0/22,152.156.0.0/16,161.0.120.0/21,164.73.0.0/16,167.56.0.0/13,167.108.0.0/16,167.116.0.0/16,168.121.184.0/22,170.247.168.0/22,179.0.156.0/22,179.24.0.0/13,181.103.64.0/20,181.224.164.0/23,186.8.0.0/18,186.8.64.0/18,186.8.128.0/17,186.48.0.0/14,186.52.0.0/14,190.0.128.0/19,190.64.0.0/17,190.64.128.0/17,190.99.96.0/20,190.108.0.0/19,190.112.52.0/22,190.112.192.0/22,190.112.200.0/22,190.132.0.0/16,190.133.0.0/16,190.134.0.0/15,192.146.237.0/24,200.0.68.0/23,200.0.86.0/23,200.0.88.0/24,200.0.204.0/22,200.2.32.0/19,200.3.12.0/22,200.7.84.0/23,200.7.86.0/23,200.10.60.0/23,200.10.62.0/23,200.12.155.0/24,200.12.156.0/24,200.35.152.0/22,200.40.0.0/17,200.40.128.0/18,200.40.192.0/19,200.40.224.0/20,200.40.240.0/20,200.58.128.0/20,200.58.144.0/20,200.71.0.0/20,200.71.16.0/20,200.93.240.0/24,200.93.241.0/24,200.93.242.0/24,200.93.243.0/24,200.93.244.0/24,200.93.245.0/24,200.93.246.0/24,200.93.247.0/24,200.108.192.0/19,200.108.224.0/20,200.108.240.0/20,200.115.72.0/21,200.124.192.0/20,200.124.208.0/20,200.125.0.0/20,200.125.16.0/20,200.125.32.0/20,200.125.48.0/20,201.217.128.0/19,201.217.160.0/19,201.219.224.0/21,201.219.252.0/22,201.221.0.0/20,201.221.16.0/20,201.221.32.0/20,201.221.48.0/20,}

Destroying temporary ipsets...
apply: Error: Failed to import the iplist from '/tmp/geoip-shell/UY_ipv4.iplist' into ip set 'UY_ipv4_2024-10-08_geoip-shell'.
manage: Warning: Discrepancy detected between the firewall state and the config file.
manage: Warning: missing ip lists in the firewall: 'UY_ipv4 UY_ipv6'
manage: Error: Failed to apply geoip-shell config.
install: Error: geoip-shell-manage.sh exited with error code 1.
```
And this is the output of the commands:
```
fanta@linux:~/firewallsito/geoip-shell-0.5.8$ uname -a
Linux linux.ypfnbctd.com 6.1.0 #1 SMP Wed Jul 12 12:00:44 MSK 2023 x86_64 GNU/Linux

fanta@linux:~/firewallsito/geoip-shell-0.5.8$ nft --version
-bash: nft: command not found

fanta@linux:~/firewallsito/geoip-shell-0.5.8$ sudo nft --version
nftables v1.0.6 (Lester Gooch #5)

fanta@linux:~/firewallsito/geoip-shell-0.5.8$ sudo apt-get install nftables
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
nftables is already the newest version (1.0.6-2+deb12u2).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
```

I tried with versions 0.5.8, 0.5.7 and 0.5.6 and is the same error on all, even changing the whitelist country.

Thanks in advance for your help

**Reply by friendly-bits** [reply]
I believe it's something in your system. Either something wrong with the kernel, or you are using some software which prevents nftables from working correctly. I just installed Debian Bookworm in a virtual machine for testing - geoip-shell works fine on it.

What's the output of:
`sudo apt list --installed linux-image*`

**Reply by Fantanchero** [reply]
```
fanta@linux:~$ sudo apt list --installed linux-image*
Listing... Done
```
This is a Ethernet Servers VPS with preinstalled Debian 12 64-bit
I'm going to try reinstalling the OS (Maybe changing the distro?)

**Reply by friendly-bits** [reply]
If you can install stock Debian, this would be best I think. If not then you could try Ubuntu.

**Reply by friendly-bits** [reply]
You could also maybe look into installing a different kernel version before complete reinstall.

**Reply by Fantanchero** [reply]
After some investigation, the error: 
`Error: Could not process rule: Numerical result out of range` 
refers to nftables maximum length of a set name is 15 characters (read 27 on other site). 
If I'm not mistaken, if the name is "UY_ipv4_2024-10-08_geoip-shell" then is 30 chars. 
Is there a way to modify the naming scheme on the script?

**Reply by friendly-bits** [reply]
Indeed, turns out that nftables wiki says the set name must be 16 characters or less:
https://wiki.nftables.org/wiki-nftables/index.php/Sets

Which is weird because geoip-shell creates sets with 31 characters-long names and this works just fine on all systems I've tested with (Debian, Linux Mint, Gentoo, OpenWrt). Perhaps there is some obscure kernel compile-time option which can be set to enforce this and most distros don't set it.

Current set names format is kinda ingrained in the geoip-shell code, so quite a bit of modification will be necessary. I'll make an experimental branch which will implement shorter set names.

**Reply by friendly-bits** [reply]
The experimental branch is here:
> https://github.com/friendly-bits/geoip-shell/tree/shorter-nftables-set-names

Please test with it and let me know if the issue is fixed.

**Reply by Fantanchero** [reply]
Working perfectly now
Thanks you very much.
```
Successfully configured geoip-shell for firewall backend: nftables.

Ip lists in the final whitelist: 'UY_ipv4 UY_ipv6'.

View geoip status with 'geoip-shell status' (may require 'sudo').

Install done.
fanta@linux:~/firewall/geoip-shell-shorter-nftables-set-names$ sudo geoip-shell status

geoip-shell status:

geoip-shell v0.5.8

Firewall backend: nftables
Geoip blocking mode: whitelist
Ip lists source: ripe
Country codes in the whitelist: UY ✔
IP families in firewall rules: ipv4 ipv6 ✔
Geoip rules applied to network interfaces: venet0

Protocols:
tcp: Geoip is applied to all ports
udp: Geoip is applied to all ports

Geoip firewall chain: enabled ✔
Whitelist blocking rule: ✔

nftables sets optimization policy: performance

Cron system service: ✔
Update cron job: ✔
Update schedule: '12 4 * * *'
Last successful update: Oct-10-2024 19:34:29
Persistence cron job: ✔
Automatic backup of ip lists: On

No problems detected.
```

**Reply by friendly-bits** [reply]
Thanks for reporting this and helping me identify the issue. The fix is now incorporated in v0.5.9.

**Comment by oraculix** [comment]
Hello Anton,
I just want to leave a big **THANK YOU!** for sharing your work here.

After some years of not having the need, I researched into Geoblocking again and – after sifting through quite some outdated stuff – luckily stumbled upon your scripts.

Within a few minutes, `geoip-shell` was up and running on my Debian Bookworm Instance that has an SSH Port open to the public Internet. And many unwelcome visitors trying to get in from day one on.

Keep up the good work!
Cheers, Uwe

**Reply by friendly-bits** [reply]
Hi Uwe and thank you for sharing your experience. This kind of responses makes it worthwhile to do open source development.

**Comment by orkaa** [comment]
Hey friendly-bits,

I also wanted to say a quick thank you!

I've been running a public endpoint for my homelab and wanted to restrict the access to my home country only. I looked into a custom nginx setup, then bunkerweb and while those worked, they only covered the HTTP(s) ports.

Then I found your project on reddit, that protects the whole server, including SSH and the setup is super simple! Loving it so far!

**Reply by friendly-bits** [reply]
Hi, thanks for this nice review. Glad geoip-shell is useful to you.

**Comment by QuaxEros** [comment]
Of course i read your welcome post!

So.. I'm since close to 3 weeks under heavy attack of botnet(s) and the guys are giving me no sleep. Using MANY thousands of single IPs and over a thousand IP-ranges. Don't know why they think it's worth their time so much...

Held up well for these weeks but my fail2ban block-lists became so large it started interfering with proper functioning of the server(s), f2b can hardly keep up and not all connects provoke errors to react on. Needed a different approach, geo-blocking seemed obvious but hard to find a decent option. Luckily stumbled upon your repo and decided to give it a go. Worried about interference with my other firewall related measures and (over)complicated setup, some NAT-guests behind hypervisor, some direct on WAN bridge, 2 web servers, 1 mail server, a panel, a docker host, internal virtual local nets, The Works... All for community and non-profit projects.

Started using Geoip-shell carefully today and for now not having any issues besides my lips cramping from my constant smile... Might even catch some sleep this night!
Having it in the MANGLE table seems to make it trouble-free pre-filtering before f2b and other rules (still have to test some though...)

Your coffee and beer is standing in front of me, come drink it with me my man !!

**Reply by friendly-bits** [reply]
Great to hear that this little project helped you with what sounds like a nightmare. I started working on it mainly to stop the bot scanning and never imagined that it would be used to stop a targeted botnet attack. I also have to say that depending on the resources and determination of the attackers, geoblocking may turn out to only provide a temporary relief, till the attackers figure it out and adapt.

About avoiding interference with your services: geoip-shell supports specifying trusted ip's or subnets which will be excluded from the filter. If having any trouble, you may need to figure out which ip's should be allowed and use that geoip-shell option. Depending on the situation, you might also need to exclude certain ports from geoblocking, which is supported by geoip-shell as well.

I appreciate the invitation for coffee/beer. However we are statistically unlikely to be even in the same geographical region (I live in a small country which I don't really want to publicly disclose). So I have to decline. I also currently don't accept donations. But I do appreciate the attitude.

Please do keep me posted on how it goes with the whole attack situation.

**Reply by QuaxEros** [reply]
I'll open up my own thread and explain my setup and use case so we can go into the specifics of how it could hinder (or not) my other defenses and rule chains. For now Asia and Brazil are the origin of all the zombies (not scattered all over the world). Yep, that is a bit strange, i feel it is almost a personal vendetta with the way the resources are used and the attack is coordinated, my servers are kind of insignificant also.. It's not a DDOS but a brute force on all services. Will have time for this in a few days i hope... CU Later..!

**Comment by namshubs** [comment]
Using openWRT i wanted to explore the use of the geoip-shell for routing only and not strictly for blocking. 

My intention is to split the traffic from my homecountry (DE) IP set directly to the WAN 
and the other Traffic i want to route through an existing openvpn tun0 dev. 

I manged such a configuration with manual ip lists some years ago but needed a lot of external help
 and a lot of shellscript workarounds.(I found i have much less patience ourdays :)

I wonder if it wouldnt be faster using the geoip-shell script for the ip-set maintainance and marking.

To access the ip-set names i would like to avoid to "extract the ip-set name with date" each time since the set name has now a date suffix.
Is there a flag or method to name the ip-set always the same ?  (IP_DE_4) or IP_DE_6)  ?

If someone has addititional recomenddations for this goal please let me know, i am quite out ouf the practise in these days. 

I admire the elegance and the smooth integration of the project! Kudos!

**Reply by friendly-bits** [reply]
Hi, thank you for the compliments. If you want to use sets created by geoip-shell for your own purposes, in principle you can configure geoip-shell with arbitrary country codes and disable geoblocking:
```
geoip-shell configure -c <"country_codes"> -N true
```

You can then use the ipsets created by geoip-shell for your purposes. As to sets naming, unfortunately sets creation logic needs the date in set names. The main reason is that when updating sets, geoip-shell creates a new set first and then removes the old ones, and you can not create a set with the same name as the existing set. This needs to be implemented this way for technical reasons which are a bit too long to explain. If you want to get a name of a particular set, you could either implement your own solution or use internal functions from the geoip-shell scripts:

```
#!/bin/sh

interesting_country_codes="<your_country_codes>"
interesting_families="4 6"

# source required geoip-shell scripts
. /usr/lib/geoip-shell-lib-common.sh &&
. /usr/lib/geoip-shell-lib-<nft|ipt>.sh ||
    { logger -t <your_script_name> -p user.err "ERROR: Failed to source geoip-shell scripts."; exit 1; }

ipsets="$(get_ipsets)" && [ -n "$ipsets" ] ||
    { logger -t <your_script_name> -p user.err "ERROR: Failed to get geoip-shell ipsets."; exit 1; }

for family in $interesting_families; do
    for country_code in $interesting_country_codes; do
        get_matching_line "$ipsets" "" "${country_code}_${family}" "*" matching_ipset

        [ -n "$matching_ipset" ] ||
            { logger -t <your_script_name> -p user.err "ERROR: Failed to get IP set name for '${country_code}_${family}'."; continue; }

        case "$matching_ipset" in *"$_nl"*
            logger -t <your_script_name> -p user.err "ERROR: Got string containing newlines for '${country_code}_${family}'."
            continue
        esac

        # do something with the ipset $matching_ipset
    done
done
```

Replace `<strings_encolsed_in_angle_brackets>`.

Note that internal geoip-shell functions API may change in future releases (although I am not planning to change these specific functions right now but I can't guarantee they won't be changed), so this script may require adjustments at some point in the future.

I can not test this code right now so it may or may not need some minor changes.

**Reply by namshubs** [reply]
Thank you for the fast reply!
First. I  understand this date suffix is part of the system.
Second, i just looked in your sample code, and i guess thats the way. 
So i will  sharpen my skills with your sample code!

**Comment by namshubs** [comment]
hello, let me know if  i stress your patience :)

to load the ip list into nftables i had at least "whitlist" the country, whithout no sets make it into nft.

e.g. 

geoip-shell configure -c "DE" -N true -m whitelist  

with a little filtering i could identify then the table names 
```nft list sets | awk '$1=="set" && $2 ~ /^RU_[0-9]+_[0-9]{4}-[0-9]{2}-[0-9]{2}$/ { print $2 }```

lists now the setnames and i can run them in my own script.
DE_4_2025-07-30
DE_6_2025-07-30

i now try to use this in the further routing script / preparation
txs meanwhile!d

**Reply by friendly-bits** [reply]
Hi, the command `nft list sets` will print sets name and sets contents, which you then need to filter. This is unnecessarily slow. Instead, you could use the command `nft -t list sets`. The `-t` option prevents sets contents from being printed.

In general, this approach has the downside of printing all sets in all tables, and if another set in another table will match the regex used in the awk command (BTW this particular regex will only match sets for Russia because of `RU`) then you get set names you did not expect. Instead you could use the command:

```
nft -t list table inet geoip-shell | awk '$1 == "set" && $2 ~ /^[A-Z][A-Z]_[46]_[0-9]{4}-[0-9]{2}-[0-9]{2}$/ {print $2}'
```

Which will print the names of all geoip-shell sets for countries.

**Reply by namshubs** [reply]
To give you a little background 
The routing solution shall be for my former coworkers and friends in Sankt Petersburg. 
Most of the foreign "western" internet is meanwhile blocked there (meta, youtube, linkedin, nearly all media outlets bbc, spiegel, welt.de etc.). Most western VPN are also not working anymore, yet a view are under the radar. 

Most of the russian websites for the dayly use (banking apps, services, railticket etc)  meanwhile detect western ips (if you switched to vpn) and geoblock them (!) . So for a normal person interested to have a global information input, is in a permanent hassel to switch vpn on and off.  
My idea is to give them a proper routing configuration with routing split via vpn tunnel or wan. That  would give the people the option at least at home to have the unblocked internet freedom back. It could work also in other restricted internet realms.
 
*actually thinking about it "is also right the opposite of what geoip-blocking is meant for... :)"
i hope you dont mind my intention. 


Regarding this nft list sets thing...

I choosed this approach since i could not get any results using the 
  ipsets="$(get_ipsets)" from your demo script. It was always empty but the nft list sets showed the table and sets

The nft list tables method was then my workaround to get the names of the tables. 
so thank you for the -t  and proper  filter.

If i get it done i will definetly publish an howto documentation as first project in GitHub, so i guess it will take a while :)
Thanks!

**Reply by friendly-bits** [reply]
Thanks for the explanation. I can see how implementing your idea could be useful. If it works out, perhaps I could consider adding a mode of operation where only sets are created and maintained without any firewall rules.

I definitely do not mind your intention. geoip-shell was created to provide a convenient open-source solution to a particular problem. If geoip-shell can accidentally help solve another problem, I'm all for it.



> i could not get any results using the
> ipsets="$(get_ipsets)" from your demo script

Then there is a bug in that sample code. As I mentioned, I could not test it when writing it. This should be easy to troubleshoot and solve but the `nft | awk` one-liner is simpler, better and does not rely on internal API, so IMO there is no need to use the sample code.



> If i get it done i will definetly publish an howto documentation as first project in GitHub

Please do let me know if/when this happens. BTW I would be happy to help you with this project, either midway or after its first public release. You can contact me privately on the OpenWrt forum.

**Reply by namshubs** [reply]
Sorry for the late reply. Short status... 
I spend the weeks free time on the task.
2 times it suddenly worked. But each time after reboot not (all traffic then again was routed with the "Catch All Rule" via vpn), the marked ips or the look up failed. Still iterate myself around the hows and whys.
Will try again after a weekend in the Nature and get back! I definetly need to learn more about tables chains and rules...

**Reply by friendly-bits** [reply]
> each time after reboot not

Note that geoip-shell sleeps for 30s after reboot, and then fetching, parsing and loading ipsets takes some more time. geoip-shell will refuse to change the config until it's done with the previous task. BTW you can always use `logread` to see what's going on during boot.
